scripts/aws_replicate.py

```diff
@@ -0,0 +1,376 @@
+
+import json
+import boto3
+
+from cdislogging import get_logger
+from indexclient.client import IndexClient
+
+from errors import APIError
+from settings import PROJECT_MAP, INDEXD
+from utils import (
+    get_bucket_name,
+    get_fileinfo_list_from_manifest,
+    get_fileinfo_list_from_s3_manifest,
+    exec_files_grouping,
+)
+
+logger = get_logger("AWSReplication")
+
+global EXIT_FLAG
+EXIT_FLAG = 0
+
+global TOTAL_PROCESSED_FILES
+TOTAL_PROCESSED_FILES = 0
+
+mutexLock = threading.Lock()
+workQueue = Queue.Queue()
+
+indexclient = IndexClient(
+    INDEXD["host"],
+    INDEXD["version"],
+    (INDEXD["auth"]["username"], INDEXD["auth"]["password"]),
+)
+
+
+class singleThread(threading.Thread):
+    def __init__(
+        self,
+        threadID,
+        threadName,
+        source_bucket,
+        copied_objects,
+        global_config,
+        job_name,
+        q,
+    ):
+        threading.Thread.__init__(self)
+        self.threadID = threadID
+        self.threadName = threadName
+        self.global_config = global_config
+        self.q = q
+        self.source_bucket = source_bucket
+        self.copied_objects = copied_objects
+        self.job_name = job_name
+        self.json_log = {}
+
+    def run(self):
+        logger.info("Starting " + self.name)
+        if self.job_name == "copying":
+            process_data(
+                self.threadName,
+                self.source_bucket,
+                self.global_config,
+                self.copied_objects,
+                self.q,
+            )
+        elif self.job_name == "indexing":
+            self.indexing_data(
+                self.threadName, self.global_config, self.copied_objects, self.q
+            )
+        logger.info("\nExiting " + self.name)
+
+    def indexing_data(self, threadName, global_config, copied_objects, q):
+        """
+        Start to index the copied object
+
+        Args:
+            threadName(str): thread name
+            global_config(dict): configuration dictionary
+            copied_objects(set): objects aready copied
+            q(Queue): queue containing the objects need to be processed
+        
+        Return:
+            None
+        """
+
+        while not EXIT_FLAG:
+            mutexLock.acquire()
+            if not q.empty():
+                files = q.get()
+                mutexLock.release()
+                logger.info("%s indexing %d files" % (threadName, len(files)))
+                check_and_index_the_data(files, copied_objects, self.json_log)
+                logger.info(self.json_log)
+
+            else:
+                s3 = boto3.client("s3")
+                self.write_log_file(s3)
+                mutexLock.release()
+
+    def write_log_file(self, s3):
+        """
+        write log file and push to s3 log bucket
+        """
+        filename = self.threadName + "_log.json"
+        with open(filename, "w") as outfile:
+            json.dump(self.json_log, outfile)
+        s3.upload_file(
```

Comment:
you can upload straight to s3 without the temporary local file by making a `boto.s3.key.Key` and setting its contents, or using StringIO/BytesIO. (unless it's desirable to have the logs save here as well)

Reply:
yeah. I also want to keep a local one